src/curvefit/uncertainty/draws.py

```diff
@@ -0,0 +1,140 @@
+
+        ## Methods
+        ### `create_draws`
+            Fills in the `draws` dictionary by generating `num_groups` sample trajectories per location
+            - `data (curvefit.core.data.Data)`: data
+            - `model_prototype (curvefit.models.base.Model)`: a model object
+            - `solver_prototype (curvefit.solvers.solver.Solver)`: a solver used to fit the model
+            - `residual_model (curvefit.uncertainty.residual_model._ResidualModel)' a residual model object, should be
+                 fitted before being passed here.
+            - `evaluation_space (callable)`: which space to generate draws in. It should be the same as used
+                in `PredictiveValidity` instance which generated the residual matrix passed to the `residual_model`
+            - `theta (float)`: between 0 and 1, how much scaling of the residuals to do relative to the prediction mean
+
+        ## `get_draws`
+            Returns generated draws
+            - `group (str)`: Group for which the draws are requested. Defaults to None, in which case returns
+                a dictionary of all draws indexed by groups.
+
+        ## Usage
+        In `ModelRunner.run()`:
+        ```python
+            draws = self.draws.create_draws(
+                data=self.data,
+                model_prototype=self.model,
+                solver_prototype=self.solver,
+                residual_model=self.residual_model,
+                evaluation_space = self.predictive_validity.evaluation_space,
+                theta=self.predictive_validity.theta
+            ).get_draws()
+            '''
+        {end_markdown _ResidualModel}
+        """
+
+    def __init__(self, num_draws, prediction_times, exp_smoothing=None, max_last=None):
+
+        self.num_draws = num_draws
+        self.prediction_times = prediction_times
+        self.exp_smoothing = exp_smoothing
+        self.max_last = max_last
+        self._draws = None
+
+        assert type(self.num_draws) == int
+        assert self.num_draws > 0
+
+        assert type(self.prediction_times) == np.ndarray
+
+        if self.exp_smoothing is not None:
+            assert type(self.exp_smoothing) == float
+            if self.max_last is None:
+                raise RuntimeError("Need to pass in how many of the last models to use.")
+            else:
+                assert type(self.max_last) == int
+                assert self.max_last > 0
+        else:
+            if self.max_last is not None:
+                raise RuntimeError("Need to pass in exponential smoothing parameter.")
+
+    # TODO: Implement exponential smoothing (do we need it?)
+    def create_draws(self,
+                     data: Data,
+                     model_prototype,
+                     solver_prototype,
+                     residual_model: _ResidualModel,
+                     evaluation_space,
+                     theta: float):
+
+        self._draws = {group: [] for group in data.groups}
```

Comment:
I think we'll want this to be an empty np.array instead of a list, because the resulting dict should be
`{group: np.array(shape=(self.num_draws, len(prediction_times))}`

Reply:
Ah, that's a bug for sure. I think it should be just `self._draws = {}` then, since noisy_forecast is already an array of `shape=(self.num_draws, len(prediction_times))`, so all we do at the end is:
`self._draws[group] = noisy_forecast`